### PR TITLE
Add extra methods to the merchant api and catagorize potions

### PR DESCRIPTION
**Pull request by frogocomics**
This is my first pull request, so I don't have high hopes for this. My would like to propose the addition of extra, but useful methods such as ```getRemainingUses()``` that would prove useful to coders. I fixed code that was not following the contribution guidelines. 

I also worked on categorizing potions into three types, ```POSITIVE```, ```NEGATIVE```, and ```UNUSED```. Lastly, I added UNUSED potions into the list of possible potions. Unused potions (Example: Bland, Gross) were implemented in early beta, and obtainable through commands.

### CODE REVIEW

**Comment by simon816** [comment]
This has merge conflicts, update your branch. (Also using your `master` branch is generally a bad idea).

File: src/main/java/org/spongepowered/api/item/merchant/TradeOfferBuilder.java [hunk 1]
```diff
@@ -73,7 +82,7 @@
      * @param maxUses The maximum uses of the trade offer
      * @return This builder
      */
-    TradeOfferBuilder maxUses(int maxUses);
+    TradeOfferBuilder setMaxUses(int maxUses);
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/item/merchant/TradeOfferBuilder.java]: In `Builder`s like this one, `set` is sorta optional as a method prefix.

File: src/main/java/org/spongepowered/api/item/merchant/Merchant.java [hunk 2]
```diff
@@ -62,6 +64,13 @@
     List<TradeOffer> getOffers();
 
     /**
+     * Gets a immutable list of Titles that were displayed to the user.
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/item/merchant/Merchant.java]: What titles? Are you talking about the title api?

File: src/main/java/org/spongepowered/api/item/merchant/TradeOfferBuilder.java [hunk 3]
```diff
@@ -87,8 +96,7 @@
      * Creates a new TradeOffer instance with the current state of the builder.
      *
      * @return A new trade offer instance
-     * @throws IllegalStateException If the resulting trade offer would be
-     *      invalid
+     * @throws IllegalStateException If the resulting trade offer would be invalid
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/item/merchant/TradeOfferBuilder.java]: Don't change this, the line is wrapped to 80 characters

File: src/main/java/org/spongepowered/api/potion/PotionEffectBuilder.java [hunk 4]
```diff
@@ -41,18 +41,18 @@
...
      * Sets the amplifier power of the potion effect.
      *
-     * <p>Amplifiers must be above zero.</p>
+     * <p>Amplifiers must be above 0 and below 255.</p>
      *
      * @param amplifier The amplifier power
-     * @return This builder, for chaining
-     * @throws IllegalArgumentException If the amplifier is less than zero
+     * @return This builder
+     * @throws IllegalArgumentException If the amplifier is less than 0 or greater than 255
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/potion/PotionEffectBuilder.java]: Need to wrap this line

File: src/main/java/org/spongepowered/api/potion/PotionEffectBuilder.java [hunk 5]
```diff
@@ -33,7 +33,7 @@
      * Sets the {@link PotionEffectType} of the potion.
      *
      * @param potionEffectType The type of item
-     * @return This builder, for chaining
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/potion/PotionEffectBuilder.java]: No reason to remove this (and others)

File: src/main/java/org/spongepowered/api/potion/PotionEffectBuilder.java [hunk 6]
```diff
@@ -68,7 +68,7 @@
      * Sets the potion effect to show particles when applied or not.
      *
      * @param showsParticles Whether the potion effect will show particles
-     * @return This builder, for chaining
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/potion/PotionEffectBuilder.java]: Why did you remove this?

File: src/main/java/org/spongepowered/api/potion/PotionTypes.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.spongepowered.api.potion;
+
+/**
+ * An enumeration of all possible potion types in vanilla minecraft.
+ */
+public final class PotionTypes {
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/potion/PotionTypes.java]: I can't suggest anything better, but `PotionTypes` implies it's a list of all the potion names (at least to me). Not so much whether it's a positive effect or a negative effect.

File: src/main/java/org/spongepowered/api/item/merchant/Merchant.java [hunk 8]
```diff
@@ -72,9 +81,37 @@
     void setOffers(List<TradeOffer> offers);
 
     /**
+     * Gets a specific offer from a list of offers.
+     *
+     * <p>Will return {@link Optional#absent()} when the index specified does not exist.</p>
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/item/merchant/Merchant.java]: Wrap this line

File: src/main/java/org/spongepowered/api/item/merchant/Merchant.java [hunk 9]
```diff
@@ -72,9 +81,37 @@
...
+     *
+     * @param index The index of the offer
+     * @return The offer
+     */
+    Optional<TradeOffer> getOffer(int index);
+
+    /**
+     * Gets a specific title from a list of titles.
+     *
+     * <p>Will return {@link Optional#absent()} when the index specified does not exist.</p>
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/item/merchant/Merchant.java]: Wrap this too

File: src/main/java/org/spongepowered/api/item/merchant/Merchant.java [hunk 10]
```diff
@@ -72,9 +81,37 @@
...
      *
      * @param offer The offer to add
      */
     void addOffer(TradeOffer offer);
+
+    /**
+     * Removes the five offer to the list of offers provided by this merchant.
+     *
+     * @param offer The offer to remove
+     * @throws NoSuchElementException The offer that is to be removed does not exist
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/item/merchant/Merchant.java]: Too long as well

File: src/main/java/org/spongepowered/api/item/merchant/TradeOfferBuilder.java [hunk 11]
```diff
@@ -25,9 +25,10 @@
 package org.spongepowered.api.item.merchant;
 
 import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.text.Text;
 
 /**
- * Represents a builder to generate immutable {@link TradeOffer}s.
+ * Represents a builder to generate {@link TradeOffer}s.
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/item/merchant/TradeOfferBuilder.java]: Why make these mutable? You don't mention the reason in your PR.

File: src/main/java/org/spongepowered/api/item/merchant/Merchant.java (hunk 10 above)
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/item/merchant/Merchant.java]: This is intended for enumerations, and shouldn't be thrown here.

File: src/main/java/org/spongepowered/api/item/merchant/Merchant.java [hunk 12]
```diff
@@ -72,9 +81,37 @@
...
+
+    /**
      * Adds the given offer to the list of offers provided by this merchant.
      *
      * @param offer The offer to add
      */
     void addOffer(TradeOffer offer);
+
+    /**
+     * Removes the five offer to the list of offers provided by this merchant.
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/item/merchant/Merchant.java]: The five offer?

File: src/main/java/org/spongepowered/api/potion/PotionTypes.java (hunk 7 above)
Comment (frogocomics) [inline src/main/java/org/spongepowered/api/potion/PotionTypes.java]: Maybe I should change the class to ```PotionCatagories```?

**Comment by Phroa** [comment]
This might be overly snobby but according to basically every style guide you should write commit messages like you're telling somebody else to write it. "Add extra methods" is more conventional.

File: src/main/java/org/spongepowered/api/item/merchant/TradeOffer.java [hunk 13]
```diff
@@ -38,6 +38,13 @@
 public interface TradeOffer {
 
     /**
+     * Get the {@link TradeOfferBuilder} used to edit trade offers.
+     *
+     * @return The trade offer builder
+     */
+    TradeOfferBuilder builder();
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/item/merchant/TradeOffer.java]: This doesn't belong here - there are already methods for getting builders in `GameRegistry`.

File: src/main/java/org/spongepowered/api/item/merchant/TradeOffer.java [hunk 14]
```diff
@@ -98,6 +105,13 @@
     int getMaxUses();
 
     /**
+     * Gets the remaining amount of uses of this offer.
+     *
+     * @return The remaining uses of this offer.
+     */
+    int getRemainingUses();
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/item/merchant/TradeOffer.java]: This can just be calculated by subtracting `getUses` from `getMaxUses`

**Comment by Aaron1011** [comment]
@frogocomics: You should split out either the potions-related changes or the TradeOffer-related changes into a separate PR. Each PR should address one topic.

**Comment by gabizou** [comment]
A few reasons why I'm closing this:
- `TradeOffer`s themselves do NOT have a "title". The "title" is the `Text` name of the `Merchant`, in almost all cases, the `Villager`
- You break all contracts of `CatalogedBy` and `CatalogType` with the unnecessary name change of `PotionEffectType` to `PotionType`
- `TradeOffer` is always going to be immutable
- We should not include unused potion effect types as they are unused and could very well disappear in future updates to Minecraft
- Useless addition of `getBuilder()` in `TradeOffer`, it simply shouldn't exist
- Lastly, your changes are all out of date since the merging of the Data API a few weeks ago

If you still want these changes, I'd recommend separating potion changes in a new PR and the `TradeOffer` changes in yet another PR.

**Comment by frogocomics** [comment]
Understood. As I'm younger than you think, I am trying the best of my ability to contribute to SpongeApi. As I'm new to GitHub, I have plenty to learn.

**Comment by gabizou** [comment]
> I am trying the best of my ability to contribute to SpongeApi.

I understand you're trying to contribute, and I've already said how you can contribute these changes in a more constructive manner:



> If you still want these changes, I'd recommend separating potion changes in a new PR and the TradeOffer changes in yet another PR.
